src/personaclient/PersonaClient.php

```diff
@@ -334,6 +334,110 @@ public function getUserByGuids($guids, $token)
         }
     }
 
+    /**
+     * Require authentication on your route
+     * @param string $sso
+     * @param string $provider
+     * @param string $statePrefix
+     * @param string $appId
+     * @access public
+     * @return void
+     */
+    public function requireAuth($sso, $provider, $statePrefix, $appId)
+    {
+        $_SESSION['loginSSO'] = $sso;
+        $_SESSION['loginAppId'] = $appId;
+        $_SESSION['loginProvider'] = $provider;
+        $_SESSION['loginStatePrefix'] = $statePrefix;
+
+        // Already authenticated
+        if($this->isLoggedIn())
+        {
+            return;
+        }
+
+        // Login
+        $this->login();
+    }
+
+    /**
+     * Authenticate a route
+     * @access public
+     * @return bool
+     */
+    public function authenticate()
+    {
+        if(isset($_REQUEST['persona:payload']))
+        {
+            $payload = json_decode(base64_decode($_REQUEST['persona:payload']),true);
+
+            if(!isset($_SESSION['loginState']) || $payload['state'] !==  $_SESSION['loginState'])
+            {
+                // Error with state - not authenticated
+                unset($_SESSION['loginState']);
+            } else
+            {
+                // Delete the login state ready for next login
+                unset($_SESSION['loginState']);
+
+                // Final step - validate the token
+                $_SESSION[$_SESSION['loginSSO']] = array(
```

Comment:
I wanted to keep the potential values in different arrays (different SSOs maybe) - but yep - it could be simpler.